uniffi/src/bindings/swift/mod.rs

```diff
@@ -0,0 +1,188 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::{
+    ffi::OsString,
+    fs,
+    fs::File,
+    io::Write,
+    path::{Path, PathBuf},
+};
+
+use anyhow::{anyhow, bail, Context, Result};
+
+pub mod gen_swift;
+pub use gen_swift::{BridgingHeader, Config, ModuleMap, SwiftWrapper};
+
+use super::super::interface::ComponentInterface;
+
+pub struct Bindings {
+    header: String,
+    library: String,
+}
+
+/// Generate uniffi component bindings for swift.
+///
+/// Unlike other target languages, binding to rust code from swift involves more than just
+/// generating a `.swift` file. We also need to produce a `.h` file with the C-level API
+/// declarations, and a `.modulemap` file to tell swift how to use it.
+///
+/// Further complicating matters is that fact that swift doesn't seem to like having multiple
+/// `.modulemap` files referring to headers in the same directory (if you try it, it will complain
+/// about multiple umbrella headers covering the same directory). That's not a problem for
+/// stand-alone crates but it doesn't work well when you have multiple uniffi components defined
+/// in the same workspace.
+///
+/// We work around this by generating the module-mapping instructions in a separate subdirectory
+/// for each module, producing a directory structure that looks like this:
```

Comment:
@linacambridge in particular could you please take a look at the yucky things I had to do here to make this work in a cargo workspace? You may be able to see a better way.

Reply:
I think we can actually get rid of the yucky stuff if we change `umbrella header` to just `header` in the `.modulemap` (that's probably technically more correct, too—"umbrella header" means "includes all the other headers in that directory", which our `-Bridging-Header.h` definitely doesn't do. There's probably other details I'm missing here; I just decided to try it on a whim).

I just tried it locally, and both `cargo test` in the top-level workspace, and `cargo run -- exec tests/bindings/test_geometry.swift` worked! 🎉